### ISSUE TITLE
Ability to create larger constants.

### ISSUE THREAD

**Issue by junderw** [cmpute/dashu#38]
There are many constant values that people might want to create in their projects that are 256 bit, 512 bit etc.

Currently with DWord we can instantiate 128 bit values on 64 bit arch.

It would be nice to be able to instantiate a const value for up to 512 bit regardless of the arch bits.

Thoughts?

**Comment by cmpute** [comment]
You can instantiate an arbitrarily large number with the `ubig!` or `ibig!` macros, although not in const context. I wish to support doing that in const context, but that's limited by the Rust, because Rust doesn't allow allocating memory in rust context yet. 

Due to the fact that dashu aims to support arbitrarily large numbers, which essentially means allocation is unavoidable, this won't be implemented until Rust supports const allocation. Thanks for your interest!

**Comment by junderw** [comment]
What about something along the lines of https://github.com/rust-num/num-bigint/pull/286 this PR (WIP) where a `Cow<'static, [T]>` is used instead of a required allocation?

Playing around with the above PR (for another bigint library) it wasn't really too bad to deal with.

**Comment by cmpute** [comment]
> What about something along the lines of https://github.com/rust-num/num-bigint/pull/286 this PR (WIP) where a `Cow<'static, [T]>` is used instead of a required allocation?
> 
> Playing around with the above PR (for another bigint library) it wasn't really too bad to deal with.

that sounds interesting for me, however I'm not sure whether it goes well with the already pretty complex internal design of dashu integers. PR is welcome!

**Comment by cmpute** [comment]
After a second thought, I think the best way to handle this is to provide a function that converts a reference to a static array to a reference of big integers using some trick. Direct instantiating a big integer is problematic because it conflicts with the internal design of ubig. I have to add a separate flag for the "borrowed" status, which improves the number of cases to be dealt with for unary and binary operators a lot.

Therefore, if something like `fn from_words_const(arr: &'static [Word]) -> &UBig` can be implemented, then that's perfect. Otherwise, I think the benefit of the const construction of big integers is not worth the making great changes to the code base.

**Comment by cmpute** [comment]
I have implemented two macros `static_ubig!` and `static_ibig!`, for creating static `UBig` and `IBig` respectively. Please let me know how you like it :) A new version will be published when I have also implemented a `static_fbig` macro.

**Comment by junderw** [comment]
What a great Christmas gift!

I'll check it out when I get back home. Thanks!

**Comment by junderw** [comment]
I have a question about https://github.com/cmpute/dashu/blob/8851ce4c8c4f03538cc6f0deffb74ff0b87ecebf/integer/src/repr.rs#L281-L282

```
This method is unsafe, because the caller must make sure that
the created instance is immutable, and drop() must not be called.
```

The argument is `&'static [Word]`

Since it's a shared reference it can not be mutable without unsafe in another place.

Since it's a `&'static` reference, it can't be dropped without unsafe in another place.

I am curious as to what possible way that someone could cause UB in safe Rust with this function.

**Comment by cmpute** [comment]
Yep, as long as the created `Repr` instance is assigned to an immutable static object, it should be fine. Merry Xmas!

**Comment by cmpute** [comment]
Unfortunately, the addition of some const functions requires Rust 1.64. Therefore I have to bump the MSRV, which requires a major version release. The new major version will take about 1 or 2 months tho..

**Comment by cmpute** [comment]
I managed to implement this without increase MSRV, so a new version is published. Please check out the `dashu::static_*` macros.
